### ISSUE TITLE
Defining a multi field in a dynamic template matching the new text type doesn't work

### ISSUE THREAD

**Issue by Bargs** [elastic/elasticsearch#16945]
**Elasticsearch version**: 5.0.0 (master)

**JVM version**: 1.8

**OS version**: OSX 10.11.3

**Description of the problem including expected versus actual behavior**:

While updating the Kibana functional tests to deal with the deprecation of the `string` mapping type, I tried to create a dynamic template matching `text` fields with a multi-field mapping as suggested in https://github.com/elastic/elasticsearch/issues/12394

My dynamic template looks like this:

```
"dynamic_templates": [{
        "text_fields": {
          "mapping": {
            "type": "text",
            "fields": {
              "raw": {
                "type": "keyword"
              }
            }
          },
          "match_mapping_type": "text",
          "match": "*"
        }
      }]
```

When posting documents with text fields to this index, the [fieldName].raw field does not get created.

**Steps to reproduce**:
 1. Create an index with the above dynamic template.
 2. Index a document that has a text field.
 3. Query the field mappings `localhost:9200/<index_name>/_mapping/*/field/*?include_defaults=true`
 4. Notice that the [fieldName].raw field is missing

I tried the same dynamic template with `"match_mapping_type": "*"` instead of `text`, and the raw field appeared as expected. I also tried the same multi field mapping on a regular field instead of a template, and that also worked. So this seems to be a problem with the combination of `text` fields and dynamic templates in particular.

**Comment by rjernst** [comment]
I think the match mapping type needs to be string? That is the "type" in json, which hasnt changed.

**Comment by rjernst** [comment]
Which raises the issue, we should validate the value of that setting for each template. We should spin a separate issue for that (probably generally about dynamic template validation as I dont think we do any right now).

**Comment by jpountz** [comment]
+1 @rjernst

**Comment by Bargs** [comment]
Confirmed `string` in the match mapping type works. That's pretty confusing for the user though. The `match_mapping_type` docs read: 

> match_mapping_type matches on the datatype detected by dynamic field mapping, in other words, the datatype that Elasticsearch thinks the field should have.

The dynamic field mapping creates `text` fields now, so I'd expect that to be a valid value. If we have to stick with string, we should at least update the match_mapping_type docs to explain this.

**Comment by pickypg** [comment]
I agree with @Bargs, plus it's inconsistent in that we _already_ allow `long` and `double` to be the matching type, which technically don't exist in JSON (which uses `number`).

**Comment by clintongormley** [comment]
@jpountz when we discussed this, I think we decided on using `text` as the detected type?

**Comment by rjernst** [comment]
We actually did use `text`, but there must still be uses of the other template methods I mentioned here: https://github.com/elastic/elasticsearch/pull/16877#discussion_r54749987

**Comment by jpountz** [comment]
@clintongormley Templates have a concept of a match type, which is used to match templates, and a dynamic type, which is the elasticsearch field type to use by default if no type is specified. When we discussed it, we indeed agreed on using text as the default dynamic type but for now we are still using "string" as the match type. I'd be fine to switch to text if we think this makes more sense, but I agree with Ryan that even more importantly we should validate templates better so that adding templates with an unknown match type would be rejected since it means those templates have no chance to be ever used.
